File: LeetCode-Java/Insertion Sort List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode insertionSortList(ListNode head) {
        if(head == null) return null;
        ListNode result = head;
        head = head.next;
        result.next = null;
        while(head!= null)
        {
            ListNode nextNode = head.next;
            ListNode preNode = null;
            ListNode tempNode = result;
            while(tempNode != null && head.val > tempNode.val)
            {
                preNode = tempNode;
                tempNode = tempNode.next;
            }
            if(preNode == null)
            {
                head.next = result;
                result = head;
            }
            else if(tempNode == null)
            {
                preNode.next = head;
                head.next = null;
            }
            else
            {
                preNode.next = head;
                head.next = tempNode;
            }
            head = nextNode;
        }
        return result;
    }
}

File: Leetcode-cpp/insertion_sort_list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* insertionSortList(ListNode* head) {
        ListNode* fakeHead = new ListNode(-1);
        while (head != nullptr) {
            ListNode* current = fakeHead;
            while (current->next != nullptr && current->next->val <= head->val) current = current->next;
            ListNode* pre_next = current->next;
            current->next = head;
            head = head->next;
            current->next->next = pre_next;
        }
        return fakeHead->next;
    }
};